File: commons-services/dictionary/src/main/java/io/github/cooperlyt/mis/service/dictionary/services/DistrictService.java
package io.github.cooperlyt.mis.service.dictionary.services;

import io.github.cooperlyt.mis.service.dictionary.Application;
import io.github.cooperlyt.mis.service.dictionary.repositories.DistrictRepository;
import io.github.cooperlyt.mis.service.dictionary.model.District;

import org.springframework.cache.annotation.CacheConfig;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.util.*;

@Service
@CacheConfig(cacheNames = "district")
public class DistrictService {

    private final DistrictRepository districtRepository;

    public DistrictService(DistrictRepository districtRepository) {
        this.districtRepository = districtRepository;
    }


    @Cacheable(key = "#level")
    public Mono<List<District>> districts(int level){
        return districtRepository.findAllByLevelGreaterThanEqual(level)
            .reduce(new District(District.ROOT_LEVEL),(root ,district) -> {
                root.putChild(district);
                return root;
            }).map(District::getChildren).cache();
    }

    public Mono<String> getAddress(int id){
        return districtRepository.getDistrictAddress(id)
            .switchIfEmpty(Mono.error(Application.ErrorDefine.DISTRICT_CODE_INVALID::exception));
    }


}

File: commons-services/dictionary/src/main/java/io/github/cooperlyt/mis/service/dictionary/services/DistrictService.kt
package io.github.cooperlyt.mis.service.dictionary.services

import io.github.cooperlyt.mis.service.dictionary.model.District
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class DistrictService(
  private val districtDao: DistrictDao
) {

  fun districts(level: Int): Mono<List<District>> {
    return districtDao.districts(level)
  }

  fun fullDistrict(id: Int): Mono<List<District>> {
    return districtDao.fullDistrict(id)
  }

  fun fullDistrictName(id: Int): Mono<List<String>> {
    return districtDao.fullDistrict(id)
      .map { it.stream().map(District::getName).toList() }
  }

  fun getAddress(id: Int): Mono<String> {
    return districtDao.getAddress(id)
  }

  fun children(code: Int): Mono<List<District>> {
    return districtDao.children(code)
  }
}